File: app/src/main/java/com/thecoolguy/rumaan/fileio/ui/activities/MainActivity.java
package com.thecoolguy.rumaan.fileio.ui.activities;

import android.Manifest;
import android.app.Dialog;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.Toast;

import com.thecoolguy.rumaan.fileio.R;
import com.thecoolguy.rumaan.fileio.data.models.LocalFile;
import com.thecoolguy.rumaan.fileio.listeners.DialogClickListener;
import com.thecoolguy.rumaan.fileio.listeners.FileLoadListener;
import com.thecoolguy.rumaan.fileio.listeners.OnFragmentInteractionListener;
import com.thecoolguy.rumaan.fileio.repository.UploadWorker;
import com.thecoolguy.rumaan.fileio.ui.fragments.ChooseFileFragment;
import com.thecoolguy.rumaan.fileio.ui.fragments.NoNetworkDialogFragment;
import com.thecoolguy.rumaan.fileio.ui.fragments.UploadFileFragment;
import com.thecoolguy.rumaan.fileio.ui.fragments.UploadProgressFragment;
import com.thecoolguy.rumaan.fileio.ui.fragments.UploadResultFragment;
import com.thecoolguy.rumaan.fileio.utils.Utils.Android;
import com.thecoolguy.rumaan.fileio.viewmodel.MainActivityViewModel;

import org.jetbrains.annotations.NotNull;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProviders;
import androidx.work.WorkStatus;
import permissions.dispatcher.NeedsPermission;
import permissions.dispatcher.OnNeverAskAgain;
import permissions.dispatcher.OnPermissionDenied;
import permissions.dispatcher.RuntimePermissions;

@RuntimePermissions
public class MainActivity extends AppCompatActivity implements DialogClickListener,
        OnFragmentInteractionListener, FileLoadListener {

    // TODO: replace all Toasts with Snackbars

    public static final String TAG = "MainActivity";
    private static final int INTENT_FILE_REQUEST = 44;

    private MainActivityViewModel viewModel;
    private ConstraintLayout rootView;
    private UploadProgressFragment progressFragment;

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.options_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_history:
                startActivity(new Intent(this, UploadHistoryActivity.class));
                return true;
            case R.id.menu_about:
                startActivity(new Intent(this, AboutActivity.class));
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == INTENT_FILE_REQUEST && resultCode == RESULT_OK) {
            Uri fileUri = data.getData();
            if (fileUri != null) {
                viewModel.chooseFileFromUri(this, fileUri);
            } else {
                Toast.makeText(this, getString(R.string.oops_some_error_occurred), Toast.LENGTH_SHORT)
                        .show();
            }
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
                                           @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        MainActivityPermissionsDispatcher
                .onRequestPermissionsResult(MainActivity.this, requestCode, grantResults);
    }

    @NeedsPermission({
            Manifest.permission.READ_EXTERNAL_STORAGE,
            Manifest.permission.WRITE_EXTERNAL_STORAGE
    })
    public void chooseFile() {
        Intent intent = Android.INSTANCE.getChooseFileIntent();
        startActivityForResult(Intent.createChooser(intent, "Choose the file to Upload.."),
                INTENT_FILE_REQUEST);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        setTheme(R.style.NoActionBarTheme);
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        rootView = findViewById(R.id.root_view);
        Toolbar toolbar = findViewById(R.id.toolbar);
        toolbar.setTitle("");
        setSupportActionBar(toolbar);

        // set up initial fragment
        setUpInitialFragment();

        viewModel = ViewModelProviders.of(this).get(MainActivityViewModel.class);
    }

    private void setUpInitialFragment() {
        getSupportFragmentManager()
                .beginTransaction()
                .add(R.id.parent_fragment_container, ChooseFileFragment.newInstance(),
                        ChooseFileFragment.TAG)
                .setTransition(FragmentTransaction.TRANSIT_FRAGMENT_CLOSE)
                .addToBackStack(ChooseFileFragment.TAG)
                .commit();
    }

    @OnPermissionDenied({
            Manifest.permission.READ_EXTERNAL_STORAGE,
            Manifest.permission.WRITE_EXTERNAL_STORAGE
    })
    void showPermissionDeniedForStorage() {
        Toast.makeText(this, getString(R.string.permission_deny), Toast.LENGTH_LONG).show();
    }

    @OnNeverAskAgain({
            Manifest.permission.READ_EXTERNAL_STORAGE,
            Manifest.permission.WRITE_EXTERNAL_STORAGE
    })
    void showAppSettings() {
        Toast.makeText(this, getString(R.string.permission_deny), Toast.LENGTH_LONG).show();
        Android.INSTANCE.showAppDetailsSettings(this);
    }

    @Override
    public void onDialogPositiveClick(@NonNull Dialog dialog, @NonNull Fragment dialogFragment) {
        if (dialogFragment instanceof NoNetworkDialogFragment) {
            Android.INSTANCE.dismissDialog(dialog);
        }
    }

    @Override
    public void onChooseFileClick() {
        MainActivityPermissionsDispatcher.chooseFileWithPermissionCheck(this);
    }

    @Override
    public void onBackPressed() {
        int backStackCount = getSupportFragmentManager().getBackStackEntryCount();
        /* Pop until first fragment i.e (ChooseFileFragment) */
        if (backStackCount > 1) {
            getSupportFragmentManager()
                    .popBackStack(ChooseFileFragment.TAG, 0);
        } else {
            finish();
        }
    }

    @Override
    public void onUploadFileClick() {
        viewModel.uploadFile();

        // showSnackBar();
        // replace the fragment container with progress

        progressFragment = UploadProgressFragment.newInstance();
        getSupportFragmentManager()
                .beginTransaction()
                .setTransition(FragmentTransaction.TRANSIT_FRAGMENT_OPEN)
                .replace(R.id.parent_fragment_container, progressFragment)
                .addToBackStack(null)
                .commit();

        if (viewModel.getUploadWorkStatus() != null) {
            viewModel.getUploadWorkStatus().observe(this, new Observer<WorkStatus>() {
                @Override
                public void onChanged(WorkStatus workStatus) {
                    if (workStatus.getState().isFinished()) {
                        String url = workStatus.getOutputData().getString(UploadWorker.KEY_RESULT);

                        // switch to results fragment
                        getSupportFragmentManager()
                                .beginTransaction()
                                .setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE)
                                .replace(R.id.parent_fragment_container, UploadResultFragment.newInstance(url),
                                        UploadResultFragment.TAG)
                                .addToBackStack(UploadResultFragment.TAG)
                                .commit();
                    }
                }
            });
        }
    }

    @Override
    public void onClose() {
        /* Pop all fragments till (Exclusive of) ChooseFileFragment from back stack */
        getSupportFragmentManager()
                .popBackStack(ChooseFileFragment.TAG, 0);
    }


    @Override
    public void onFileLoad(@NotNull LocalFile localFile) {
        // change the current fragment to upload
        FragmentManager fragmentManager = getSupportFragmentManager();
        FragmentTransaction transaction = fragmentManager.beginTransaction();
        transaction.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_OPEN);
        transaction.replace(R.id.parent_fragment_container,
                UploadFileFragment.newInstance(localFile.getName()),
                UploadFileFragment.TAG);
        transaction.addToBackStack(UploadFileFragment.TAG);
        transaction.commit();


    }
}

File: app/src/main/java/com/thecoolguy/rumaan/fileio/ui/activities/MainActivity.kt
package com.thecoolguy.rumaan.fileio.ui.activities

import android.Manifest
import android.app.Activity
import android.app.Dialog
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.work.WorkManager
import com.google.android.material.snackbar.Snackbar
import com.thecoolguy.rumaan.fileio.R
import com.thecoolguy.rumaan.fileio.listeners.DialogClickListener
import com.thecoolguy.rumaan.fileio.listeners.OnFragmentInteractionListener
import com.thecoolguy.rumaan.fileio.utils.createUploadWork
import com.thecoolguy.rumaan.fileio.repository.UploadWorker
import com.thecoolguy.rumaan.fileio.ui.fragments.HomeFragment
import com.thecoolguy.rumaan.fileio.ui.fragments.NoNetworkDialogFragment
import com.thecoolguy.rumaan.fileio.ui.fragments.ResultFragment
import com.thecoolguy.rumaan.fileio.utils.Utils
import com.thecoolguy.rumaan.fileio.utils.Utils.Android
import com.thecoolguy.rumaan.fileio.utils.toggleClickable
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.layout_main.*
import permissions.dispatcher.NeedsPermission
import permissions.dispatcher.OnNeverAskAgain
import permissions.dispatcher.OnPermissionDenied
import permissions.dispatcher.RuntimePermissions
import timber.log.Timber

@RuntimePermissions
class MainActivity : AppCompatActivity(), OnFragmentInteractionListener, DialogClickListener {
    override fun onDialogPositiveClick(dialog: Dialog, dialogFragment: Fragment) {
        dialog.dismiss()
    }

    override fun onDone() {
        this.onBackPressed()
    }

    override fun onUploadFileClick() {
        if (Utils.Android.isConnectedToNetwork(this)) {
            btn_choose_file toggleClickable false
            chooseFileWithPermissionCheck()
        } else {
            NoNetworkDialogFragment.newInstance()
                    .show(supportFragmentManager, getString(R.string.no_net_dialog_fragment_tag))
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.options_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.menu_history -> {
                startActivity(Intent(this, UploadHistoryActivity::class.java))
                return true
            }
            R.id.menu_about -> {
                startActivity(Intent(this, AboutActivity::class.java))
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    private fun setUpWork(uri: Uri) {
        val work = createUploadWork(uri.toString())
        WorkManager.getInstance().enqueue(work)

        Toast.makeText(this, "Uploading..", Snackbar.LENGTH_SHORT).show()

        val workId = work.id
        WorkManager.getInstance().getWorkInfoByIdLiveData(workId)
                .observe(this@MainActivity, Observer { workInfo ->
                    if (workInfo != null && workInfo.state.isFinished) {
                        val outputData = workInfo.outputData
                        val url = outputData.getString(UploadWorker.KEY_RESULT_URL) ?: ""
                        val days = outputData.getInt(UploadWorker.KEY_RESULT_DAYS, 14)
                        showResult(url, days)
                    }
                })
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, intent: Intent?) {
        super.onActivityResult(requestCode, resultCode, intent)
        if (requestCode == FILE_REQUEST) {
            btn_choose_file toggleClickable true
            /* Check the returned URI */
            if (resultCode == Activity.RESULT_OK) {
                intent?.data?.let { it ->
                    setUpWork(it)
                }
            }
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>,
                                            grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        onRequestPermissionsResult(requestCode, grantResults)
    }

    @NeedsPermission(Manifest.permission.READ_EXTERNAL_STORAGE, Manifest.permission.WRITE_EXTERNAL_STORAGE)
    fun chooseFile() {
        val intent = Android.chooseFileIntent
        startActivityForResult(intent, FILE_REQUEST)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        setTheme(R.style.NoActionBarTheme)
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        findViewById<Toolbar>(R.id.toolbar).apply {
            title = ""
            setSupportActionBar(this)
        }

        initFragment()
    }

    private fun initFragment() {
        // Initialize the Main Fragment layout here
        val homeFragment = HomeFragment.newInstance()

        supportFragmentManager.beginTransaction()
                .addToBackStack(HomeFragment.TAG)
                .add(container.id, homeFragment)
                .commit()

    }

    private fun showResult(url: String, days: Int) {
        val resultFragment = ResultFragment.newInstance(url, days)
        supportFragmentManager.beginTransaction()
                .addToBackStack(ResultFragment.TAG)
                .add(container.id, resultFragment)
                .commit()
    }

    override fun onBackPressed() {
        Timber.d("Back Stack count: ${supportFragmentManager.backStackEntryCount}")
        if (supportFragmentManager.backStackEntryCount > 1) {
            supportFragmentManager.popBackStack(HomeFragment.TAG, 0)
        } else {
            finishAffinity()
        }
    }

    @OnPermissionDenied(Manifest.permission.READ_EXTERNAL_STORAGE, Manifest.permission.WRITE_EXTERNAL_STORAGE)
    internal fun showPermissionDeniedForStorage() {
        Toast.makeText(this, getString(R.string.permission_deny), Toast.LENGTH_LONG).show()
    }

    @OnNeverAskAgain(Manifest.permission.READ_EXTERNAL_STORAGE, Manifest.permission.WRITE_EXTERNAL_STORAGE)
    internal fun showAppSettings() {
        Toast.makeText(this, getString(R.string.permission_deny), Toast.LENGTH_LONG).show()
        Android.showAppDetailsSettings(this)
    }

    companion object {
        const val TAG = "MainActivity"
        private const val FILE_REQUEST = 44
    }
}